### ISSUE TITLE
Add PDO::ATTR_EMULATE_PREPARES = true support

### ISSUE THREAD

**Issue by Zlob** [laravel/ideas#2005]
Hi!

Many projects have to use connection poolers like pg_bounce. So, PDO::ATTR_EMULATE_PREPARES connection param is need to be set to true. Unfortunately, Laravel does not work properly with this option:
[#29023](https://github.com/laravel/framework/issues/29023)
[#23850](https://github.com/laravel/framework/issues/23850)
[#25818](https://github.com/laravel/framework/issues/25818)
[#27951](https://github.com/laravel/framework/issues/27951)
[#28149](https://github.com/laravel/framework/issues/28149)

The problem is in bindValues and prepareBindings methods.

I see two ways to fix this problem:
1. Extend [PostgresConnection](https://github.com/illuminate/database/blob/master/PostgresConnection.php), or base [Connection](https://github.com/illuminate/database/blob/master/Connection.php) classes so they can work with disabled prepared statements
2.Or extract binding strategy into separated class, which can be easily overridden in user land.

What do you think?

**Comment by mfn** [comment]
> poolers like pg_bounce. So, PDO::ATTR_EMULATE_PREPARES connection param is need to be set to true

See my comment https://github.com/laravel/framework/issues/29023#issuecomment-507647693

You can definitely operate with pgboncer and _not_ using ATTR_EMULATE_PREPARES.

Not sure exactly, but this _may_ depend on the `pool_mode`? I use `session` mostly. If so, this should be mentioned explicitly somewhere.

**Comment by Zlob** [comment]
You are right, but for performance reasons we use transaction mode, which means we have to disable prepared statements.
I think, there might be many reasons why prepared statements must be disabled debends on paricular project. The problem is that for now Laravel do not support this out of the box and do not provides an easy way to fix it. I'd be happy make a PR for that, but I need to know wich way of fixing this problem is more acceptable.

**Comment by Vanekru** [comment]
I have similar problems. I can't affect the database architecture and pgbouncer mode. @Zlob how do you solve this problem?

**Comment by Zlob** [comment]
@Vanekru We just overrode PostgresConnection behavior in our extension package https://github.com/umbrellio/laravel-pg-extensions/commit/5f729da4d9aa039586d433e9735ac129596875af

**Comment by owengo** [comment]
Hello, I do not know how to plus on this, but this is a blocking issue for us. Currently we cannot use a postgresql backend because of this.

**Comment by owengo** [comment]
> You can definitely operate with pgboncer and _not_ using ATTR_EMULATE_PREPARES.
> 
> Not sure exactly, but this _may_ depend on the `pool_mode`? I use `session` mostly. If so, this should be mentioned explicitly somewhere.

No you can't. The "session" pooling is not really a pooling, you have one backend connexion for every connected client.
The "transaction" pooling shares a ( small ) pool of backend connexions with a big number of clients.
The idea is that the number of backend connexions at any time is the number of running transactions. So we can have hundreds of connected clients ( for example containers running in k8s pods .. ) with only a few  backend connexions.

In postgresql it is rather expensive to provision a huge number of backend connexions ( contrary to mysql which is multithreaded, each connexion is a forked process in postgresql, and also the authorization processing is not very fast ).
So it is very, very standard to put the server behind a connexion pooler ( pgbouncer is one of them ).

**Comment by mfn** [comment]
FWIW I'm using session pool with Laravel in production without ATTR_EMULATE_PREPARES.

But, I think we very different requirements. My impression is that yours has to be much much higher then e.g. mine.

I certainly see benefits of the way I operate pgbouncer but I can confirm it is not on the level of client connections <> pool connection you hint at!

**Comment by owengo** [comment]
What happens with transaction pooling is that backend sessions are shared between clients.
So any action which will alter the state of backend session will break something: SET instructions, temporary tables, etc.  When you try to alter the session in a request - for example SET var=value - there is no guarantee that your next request will be executed on the same backend session ( unless both requests are done in the same transaction ). Also your SET action will affect other queries by other clients.
Prepared statements  are temporary stored procedures, which exist only at the session level. So when PDO creates a prepared statements,  there is chance that the call of the procedure falls on a  different session where the statement does not exist. There is also a chance that the statement is there but it is another one!.
"session pooling" completely prevents these problems because there is a 1 <=> 1 mapping between clients and backend sessions, and the backend sessions are cleaned up when a client disconnects, so that the backend session can be used by a new client. 
Indeed, session pooling works very well as long as you have a small and controllable number of client connexion. But when your pool is full ( ie: when there are as many clients connected as the pool allows ), the other clients are blocked until one existing client disconnects and so releases a pool connexion. So typically when you have a lot of idle clients they will fill the pool and prevent new connexions, so your applications will not be able to connect to the database. This means that you can completely break all your applications just because your pooler is full ( for example some new application was deployed and it just initiates 20 database connexions at startup in order to provision its internal pool ).

With transaction pooling you can configure a huge number of client connexions, several thousands if you want, it will have 0 impact on performance,  0 impact on the backend. It is the number of concurrent transactions which is the limiting factor, so it is related to the capacity of your server.
